runtime/lua/vim/treesitter.lua

```diff
@@ -153,7 +153,7 @@ end
 
 --- Returns the node's range or an unpacked range table
 ---
----@param node_or_range (TSNode|table) Node or table of positions
+---@param node_or_range (TSNode | table) Node or table of positions
```

Comment:
can the generator do this workaround when generated the foo.txt vimdoc? seems like this will be a common issue.

Reply:
@lewis6991 (who added the `TSNode` -> `|TSNode|` transformation that introduced the parsing error)

Reply:
If it's easier we can remove the tag chars around `TSNode`.

Reply:
Those links are a bit helpful, though -- and I did a similar thing manually for `LanguageTree`, which makes me wonder whether it's worth looking at a more generic approach for "Lua objects" that can be made robust?

Reply:
How about we add an alternate way of referencing tags that doesn't conflict with emmylua? `|` is problematic anyway since you don't know whether `|` begins or ends a ref.

We could have something like `<|tsnode|>`?

Or we could escape any `|` that's not used for a tag ref? That'll probably be the easier.

So the docgen would take:

```
---@param node_or_range (TSNode|table) Node or table of positions
```

and output:

```
---@param node_or_range (|TSNode|\|table) Node or table of positions
```

Reply:
There's a conflict in formats (emmylua & vimdoc) that needs to be resolved. I think escaping characters is fine here as it forces vimdoc to interpret `|` literally.

Reply:
> But also, it's our format. We can do anything we like with it if it enables better docs.

That's not strictly true; it's shared with Vim, where we take a huge chunk of the docs from. And having multiple ways to do the same thing is exactly what we want to get away from.

So I have to admit I'm not crazy about that idea. The alternative is to use something different than `|` for optional/alternative arguments in the docs (and convert in docgen).

Reply:
> and output:
> 
> ```
> ---@param node_or_range (|TSNode|\|table) Node or table of positions
> ```

can it "escape" by surrounding with spaces instead?

    ---@param node_or_range (|TSNode| | table) Node or table of positions